### ISSUE TITLE
Brakeman Integration

### ISSUE THREAD

**Issue by jtmkrueger** [w0rp/ale#385]
I run [brakeman](https://github.com/presidentbeef/brakeman) regularly and it would be amazing if it could work with ale. I'm not sure if static analysis is within the scope of ale though, so this may be a non-starter.

**Comment by w0rp** [comment]
If someone wants to add this, go for it. This might require me implementing #333 first.

**Comment by jtmkrueger** [comment]
I'll read through a few of the other linters and see if I can put something together. Good to know that ale is going to be allowed to do static analysis! Would the best idea be to work off that `only-run-on-save`  branch?

**Comment by w0rp** [comment]
@jtmkrueger I'd ignore that branch. I might start again on it.

**Comment by jtmkrueger** [comment]
Cool @w0rp I'll just work off master then.

**Comment by elebow** [comment]
The tricky part about brakeman is that not only does the file need to be stored on disk, but the checker needs to be passed the path of (or invoked from) the root of a Rails application.

@w0rp how do you feel about an approach like the following for detecting the root of the Rails app?

```vim
function! ale_linters#ruby#brakeman#GetCommand(buffer) abort
  " Find the nearest dir contining "app", "db", and "config", and assume it is
  " the root of a Rails app.
  let l:app_parent    = fnamemodify(ale#path#FindNearestDirectory(a:buffer, "app"), ":p")
  let l:db_parent     = fnamemodify(ale#path#FindNearestDirectory(a:buffer, "db"), ":p")
  let l:config_parent = fnamemodify(ale#path#FindNearestDirectory(a:buffer, "config"), ":p")

  if (l:app_parent != l:db_parent) || (l:db_parent != l:config_parent)
    return ''
  end

  let l:rails_root = l:app_parent

  " This is incorrect, since a valid Rails app could exist at the filesystem root.
  " But I think it's a safe assumption for the vast majority.
  if l:rails_root == '/'
    return ''
  end

  return 'brakeman -f json -p ' . l:rails_root
endfunction
```

**Comment by w0rp** [comment]
This general kind of approach can be okay, but you will need to replace the `systemlist` call with with the `command_chain` feature. See `:help ale-command-chain`. (I just added the tag for that right _now_).

**Comment by w0rp** [comment]
Actually, you shouldn't make any calls to `find`. It won't work on Windows, and it will probably mess up on Mac OSX, because the BSD tools are crazy. Have a look at functions like `ale#path#FindNearestFile` and friends. They might help you.

**Comment by elebow** [comment]
@jtmkrueger  Did you want to work on this? I don't want to step on your toes, or duplicate work if you've already started.

**Comment by jtmkrueger** [comment]
I haven't had time @elebow and didn't get very far at all. Feel free to put something together, I won't be offended at all 😄

**Comment by elebow** [comment]
This can be closed since #509. I forgot to update that PR title before it was merged.
